kmod/patch/livepatch-patch-hook.c

```diff
@@ -0,0 +1,324 @@
+				reloc++;
+			}
+			object++;
+		}
+	}
+}
+#endif
+
+extern struct kpatch_patch_func __kpatch_funcs[], __kpatch_funcs_end[];
+extern struct kpatch_patch_dynrela __kpatch_dynrelas[], __kpatch_dynrelas_end[];
+
+static int __init patch_init(void)
+{
+	struct kpatch_patch_func *kfunc;
+	struct kpatch_patch_dynrela *dynrela;
+	struct klp_object *lobjects, *lobject;
+	struct klp_func *lfuncs, *lfunc;
+	struct klp_reloc *lrelocs, *lreloc;
+	struct patch_object *object;
+	struct patch_func *func;
+	struct patch_reloc *reloc;
+	int ret = 0, i, j;
+
+	/* organize functions and relocs by object in scaffold */
+	for (kfunc = __kpatch_funcs;
+	     kfunc != __kpatch_funcs_end;
+	     kfunc++) {
+		ret = patch_add_func_to_object(kfunc);
+		if (ret)
+			goto out;
+	}
+
+	for (dynrela = __kpatch_dynrelas;
+	     dynrela != __kpatch_dynrelas_end;
+	     dynrela++) {
+		ret = patch_add_reloc_to_object(dynrela);
+		if (ret)
+			goto out;
+	}
+
+	/* past this point, only possible return code is -ENOMEM */
+	ret = -ENOMEM;
+
+	/* allocate and fill livepatch structures */
+	patch = kzalloc(sizeof(*patch), GFP_KERNEL);
+	if (!patch)
+		goto out;
+
+	lobjects = kzalloc(sizeof(*lobjects) * (patch_objects_nr+1),
+	                      GFP_KERNEL);
+	if (!lobjects)
+		goto out;
+	patch->mod = THIS_MODULE;
+	patch->objs = lobjects;
+
+	i = 0;
+	list_for_each_entry(object, &patch_objects, list) {
+		lobject = &lobjects[i];
+		lobject->name = object->name;
+		lfuncs = kzalloc(sizeof(struct klp_func) *
+		                 (object->funcs_nr+1), GFP_KERNEL);
+		if (!lfuncs)
+			goto out;
+		lobject->funcs = lfuncs;
+		j = 0;
+		list_for_each_entry(func, &object->funcs, list) {
+			lfunc = &lfuncs[j];
+			lfunc->old_name = func->kfunc->name;
+			lfunc->new_func = (void *)func->kfunc->new_addr;
+			lfunc->old_addr = func->kfunc->old_addr;
+			j++;
+		}
+
+		lrelocs = kzalloc(sizeof(struct klp_reloc) *
+		                    (object->relocs_nr+1), GFP_KERNEL);
+		if (!lrelocs)
+			goto out;
+		lobject->relocs = lrelocs;
+		j = 0;
+		list_for_each_entry(reloc, &object->relocs, list) {
+			lreloc = &lrelocs[j];
+			lreloc->loc = reloc ->kdynrela->dest;
+			lreloc->val = reloc->kdynrela->src;
+			lreloc->type = reloc->kdynrela->type;
+			lreloc->name = reloc->kdynrela->name;
+			lreloc->addend = reloc->kdynrela->addend;
+			lreloc->external = reloc->kdynrela->external;
+			j++;
+		}
+
+		i++;
+	}
+
+	/*
+	 * Once the patch structure that the live patching API expects
+	 * has been built, we can release the scaffold structure.
+	 */
+	patch_free_scaffold();
+
+	ret = klp_register_patch(patch);
+	if (ret) {
+		patch_free_livepatch(patch);
+		return ret;
+	}
+
+	ret = klp_enable_patch(patch);
+	if (ret) {
+		WARN_ON(klp_unregister_patch(patch));
+		patch_free_livepatch(patch);
+		return ret;
+	}
+
+	return 0;
+out:
+	patch_free_livepatch(patch);
+	patch_free_scaffold();
+	return ret;
+}
+
+static void __exit patch_exit(void)
+{
+	WARN_ON(klp_disable_patch(patch));
```

Comment:
We shouldn't disable here.  That will be done by the user (which in the future will result in a module_put on this module, allowing this exit routine to be called).